frontend/src/containers/ConfigStep/DateRangePicker/DateRangePickerGroup.tsx

```diff
@@ -6,20 +11,111 @@ import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
+};
+
+type Props = {
+  sortType: SortType;
+  onChange?: (data: ISortedDateRangeType[]) => void;
+  onError?: (data: ISortedDateRangeType[]) => void;
+};
+
+const fillDateRangeGroup = <T,>(item: T, index: number) => ({
+  ...item,
+  startDateError: null,
+  endDateError: null,
+  sortIndex: index,
+});
+
+export const DateRangePickerGroup = ({ sortType, onError }: Props) => {
   const dispatch = useAppDispatch();
   const dateRangeGroup = useAppSelector(selectDateRange);
   const isAddButtonDisabled = dateRangeGroup.length === MAX_TIME_RANGE_AMOUNT;
+  const [sortedDateRangeList, setSortedDateRangeList] = useState<ISortedDateRangeType[]>(
+    dateRangeGroup.map(fillDateRangeGroup),
+  );
+
+  useEffect(() => {
+    const errors = sortedDateRangeList.filter(({ startDateError, endDateError }) => startDateError || endDateError);
+    onError?.(errors);
+  }, [onError, sortedDateRangeList]);
+
+  const handleError = (type: string, error: DateValidationError, index: number) => {
+    setSortedDateRangeList(
+      sortedDateRangeList.map((item) => ({ ...item, [type]: item.sortIndex === index ? error : null })),
+    );
+  };
+
+  const dispatchUpdateConfig = () => {
+    dispatch(updateShouldGetBoardConfig(true));
+    dispatch(updateShouldGetPipelineConfig(true));
+    dispatch(initDeploymentFrequencySettings());
+  };
+
   const addRangeHandler = () => {
-    const newDateRangeGroup = [...dateRangeGroup, { startDate: null, endDate: null }];
-    dispatch(updateDateRange(newDateRangeGroup));
+    const result = [...sortedDateRangeList, { startDate: null, endDate: null }];
+    setSortedDateRangeList(result.map(fillDateRangeGroup));
+    dispatch(updateDateRange(result.map(({ startDate, endDate }) => ({ startDate, endDate }))));
   };
+
+  const handleChange = (
+    { startDate, endDate }: { startDate: string | null; endDate: string | null },
+    index: number,
+  ) => {
+    const result = sortedDateRangeList.map((item) =>
+      item.sortIndex === index ? { ...item, startDate, endDate, startDateError: null, endDateError: null } : item,
+    );
+    setSortedDateRangeList(result);
+    dispatchUpdateConfig();
+    dispatch(updateDateRange(result.map(({ startDate, endDate }) => ({ startDate, endDate }))));
+  };
+
+  const handleRemove = (index: number) => {
+    const result = [...sortedDateRangeList];
+    remove(result, ({ sortIndex }) => sortIndex === index);
+    setSortedDateRangeList(result);
+    dispatchUpdateConfig();
+    dispatch(updateDateRange(result.map(({ startDate, endDate }) => ({ startDate, endDate }))));
+  };
+
   return (
     <DateRangePickerGroupContainer>
       <LocalizationProvider dateAdapter={AdapterDayjs}>
-        {dateRangeGroup.map(({ startDate, endDate }, index) => (
-          <DateRangePicker startDate={startDate} endDate={endDate} index={index} key={index} />
+        {sortBy(sortedDateRangeList, get(sortFn, sortType)).map(({ startDate, endDate, sortIndex }, index) => (
```

Comment:
move this sort logic into redux, keep this component easy, just get the data to render, not contain much sort logic into the component -> decouple the UI and data

Reply:
The reasons for doing this are as follows:

We have added new UI states here, which are not shared states and are not suitable for being managed in Redux.
Clearly, placing them in state offers better performance; these states are data that needs to be updated repeatedly. Whether considering performance or logical readability, it is more advisable to manage them in state.
Pure persistent data should be stored in Redux. This data will run throughout the lifecycle of the app, rather than being confined to a single component.

Please refer to the 'Should I put form state or other UI state in my store?' section of the official[ Redux documentation.](https://redux.js.org/faq/organizing-state#organizing-state-only-redux-state)